<a href="https://colab.research.google.com/github/tomashphill/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

1. Null Hypothesis:
Address and Travel Time are INDEPENDENT of each other, they are NOT statistically correlated.
2. Alternative Hypothesis:
Address and Travel Time are DEPENDENT of each other, they ARE statistically correlated.
3. Confidence Level:
95%
Conclusion:
Based on a T-Test statistic of 43.9 and a p-value off 0, we reject the null hypothesis that address (rural vs urban) and travel time are independent and suggest the alternative that there is a strong relationship between these two categorical variables.
This makes sense as students who live in rural areas would be expected to have to travel further to attend school. This matches our intuitions.

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
### YOUR WORK HERE
import pandas as pd
import numpy as np

columns= ['symboling', 'normalized-losses', 'make', 'fuel-type',
          'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
          'engine-location', 'wheel-base', 'length', 'width', 'height',
          'curb-height', 'engine-type', 'num-of-cylinders', 'engine-size',
          'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepwoer',
          'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

# Read in csv file, adding column names
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=columns)
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [ ]:
# Replace '?' with NaN
df = df.replace('?', np.NaN)
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

## Testing: Is the Sample Reflective of the Population Value in Regards to `highway-mpg`?

---

#### 💚 Null Hypothesis 💚
The sample is reflective of the population value.

---

#### ❤️ Alternative Hypothesis ❤️
The sample is **not** reflective of the population value.

---

#### 💛 Confidence Level 💛

95%

---

In [ ]:
### YOUR WORK HERE
from scipy import stats

# Pick random sample of n=10 cars
sample = df.sample(n=10, random_state=30)

# 1-sample t-test for highway-mpg
# first, get population mean for highway-mpg
population_mean = df['highway-mpg'].mean()
stats.ttest_1samp(sample['highway-mpg'], population_mean)

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

---

#### 💜 Conclusion 💜

Based on a T-Test statistic of 0.89 and a p-value off 0.40, we fail to reject the null hypothesis that the sample of cars reflects the population of cars.

---

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

## Testing: Does the sample reflect the salesman's claim that his cars typically have a fuel efficancy of about 35 miles per hour?

---

#### 💚 Null Hypothesis 💚

The sample is reflective of the salesman's claim.

---

#### ❤️ Alternative Hypothesis ❤️

The sample is **not** reflective of the salesman's claim.

---

#### 💛 Confidence Level 💛

95%

---

In [ ]:
### YOUR WORK HERE
# 1-sample t-test for highway-mpg
claim = 35
stats.ttest_1samp(sample['highway-mpg'], claim)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

---

#### 💜 Conclusion 💜

Based on a T-Test statistic of -0.96 and a p-value off 0.36, we fail to reject the null hypothesis, or the salesman's claim that his cars typically have a fuel efficiancy of about 35 miles per hour.

---

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
### YOUR WORK HERE
# new sample of 100 cars
sample = df.sample(n=100, random_state=30)

# 1-sample t-test for highway-mpg
stats.ttest_1samp(sample['highway-mpg'], claim)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

---

#### 💜 Conclusion 💜

Based on a T-Test statistic of -6.52 and a p-value off 0.000000003, we reject the null hypothesis, or the salesman's claim that his cars typically have a fuel efficiancy of about 35 miles per hour. This is a different conclusion than the one of the sample size of 10.

---

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?
> The larger sample size is more reflective of the population due to its size. It has more precision than the smaller sample. The smaller sample did not accurately reflect the population's mean value as well as the larger one did.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE
# get column names
columns = ['party','handicapped-infants','water-project',
           'budget','physician-fee-freeze', 'el-salvador-aid',
           'religious-groups','anti-satellite-ban',
           'aid-to-contras','mx-missile','immigration',
           'synfuels', 'education', 'right-to-sue','crime','duty-free',
           'south-africa']

# read in data
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=columns)

# replace answers with integers
df = df.replace({'y':1, 'n':0, '?':np.NaN})
df.head(10)

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0
5    democrat                  0.0  ...        1.0           1.0
6    democrat                  0.0  ...        1.0           1.0
7  republican                  0.0  ...        NaN           1.0
8  republican                  0.0  ...        0.0           1.0
9    democrat                  1.0  ...        NaN           NaN

[10 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [ ]:
### YOUR WORK HERE

# separate dataframe into Democrats and Republicans
rep = df.loc[lambda df: df['party'] == 'republican']
dem = df.loc[lambda df: df['party'] == 'democrat']

## Testing: Support for the `religious-groups` bill

---

#### 💚 Null Hypothesis 💚

Republican support for the `religious-groups` bill is equal to Democrat support of the `religious-groups` bill

---

#### ❤️ Alternative Hypothesis ❤️

Republican support for the `religious-groups` bill is **not** equal to Democrat support of the `religious-groups` bill

---

#### 💛 Confidence Level 💛

99%

---

In [ ]:
# Perform 2 sample t-test on republican and democratic support of religious-groups bill
stats.ttest_ind(rep['religious-groups'], dem['religious-groups'], nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

❗️*t-test statistic is positive, indicating that Republican mean is higher than Democratic mean*

---

#### 💜 Conclusion 💜

Based on a T-Test statistic of 9.74 and a p-value off 2.39e-20, we reject the null hypothesis that Republican and Democratic support for the `religious-groups` bill is equal. It turns out that Republicans support the bill more than Democrats do.

---

## Testing: Support for the `budget` bill

---

#### 💚 Null Hypothesis 💚

Republican support for the `budget` bill is equal to Democrat support of the `budget` bill

---

#### ❤️ Alternative Hypothesis ❤️

Republican support for the `budget` bill is **not** equal to Democrat support of the `budget` bill

---

#### 💛 Confidence Level 💛

99%

---

In [ ]:
# Perform 2 sample t-test on republican and democratic support of budget bill
stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

❗️*t-test statistic is positive, indicating that Republican mean is higher than Democratic mean*

---

#### 💜 Conclusion 💜

Based on a T-Test statistic of -23.21 and a p-value off -2.17e-77, we reject the null hypothesis that Republican and Democratic support for the `budget` bill is equal. It turns out that Republicans support the bill much less than Democrats do.

---

## Testing: Support for the `water-project` bill

---

#### 💚 Null Hypothesis 💚

Republican support for the `water-project` bill is equal to Democrat support of the `water-project` bill

---

#### ❤️ Alternative Hypothesis ❤️

Republican support for the `water-project` bill is **not** equal to Democrat support of the `water-project` bill

---

#### 💛 Confidence Level 💛

90%

---

In [ ]:
# Perform 2 sample t-test on republican and democratic support of water-project bill
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

---

#### 💜 Conclusion 💜
Based on a T-Test statistic of 0.08 and a p-value off 0.93, we fail to reject the null hypothesis that Republican and Democratic support for the `water-project` bill is equal. With a 90% confidence level, Democrats and Republicans did have equal support for the bill.

---

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE
# Dictionary comprehension on 2-sample tests for each bill
{bill:stats.ttest_ind(rep[bill], dem[bill], nan_policy='omit') for bill in df if bill != 'party'}

{'aid-to-contras': Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54),
 'anti-satellite-ban': Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31),
 'budget': Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77),
 'crime': Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47),
 'duty-free': Ttest_indResult(statistic=-12.853146132542978, pvalue=5.997697174347365e-32),
 'education': Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64),
 'el-salvador-aid': Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68),
 'handicapped-infants': Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18),
 'immigration': Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066),
 'mx-missile': Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47),
 'physician-fee-freeze': Ttest_indResult(statistic=49.367081573



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





![alt text](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png)

In [ ]:
### YOUR WORK HERE
import math

# made up data
example = pd.Series([1,2,3,4,5,6,7,8,9,10])

# Pass in the data and let the function find mean, std, p-value
def ttest_1sample(sample_data, hypothesized_mean):
  t_stat = test_statistic(sample_data.mean(), hypothesized_mean, sample_data.std(),len(sample))
  return t_stat
  
# Written out equation from image above
def test_statistic(sample_mean, population_mean, sample_standard_deviation, sample_size):
  return (sample_mean - population_mean) / (sample_standard_deviation / math.sqrt(sample_size))

ttest_1sample(example, 6)


-1.6514456476895407

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE